### PR TITLE
Implement SeekableAlluxioFileOutStream

### PR DESCRIPTION
**Pull request by apc999**
### What changes are proposed in this pull request?

This is an experimental feature to support seek in AlluxioFileOutStream 

### Why are the changes needed?

An experiment for  https://github.com/Alluxio/alluxio/pull/14880

### CODE REVIEW

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 1]
```diff
@@ -216,10 +234,13 @@ private int createInternal(String path, long mode, FuseFileInfo fi) {
...
     }
     try {
-      FileOutStream os = mFileSystem.createFile(uri,
-          CreateFilePOptions.newBuilder()
-              .setMode(new Mode((short) mode).toProto())
-              .build());
+      CreateFilePOptions.Builder optionBuilder = CreateFilePOptions.newBuilder()
+          .setMode(new Mode((short) mode).toProto());
+      if (mWriteThroughFilePattern != null &&
+          mWriteThroughFilePattern.matcher(path).matches()) {
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:240]: Some performance overhead when using this approach?

File: tests/src/test/java/alluxio/client/fs/io/SeekableAlluxioFileOutStreamIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+import alluxio.client.WriteType;
+import alluxio.client.file.FileOutStream;
+import alluxio.client.file.SeekableAlluxioFileOutStream;
+import alluxio.grpc.CreateFilePOptions;
+import alluxio.util.io.PathUtils;
+
+/**
+ * Integration tests for {@code SeekableAlluxioFileOutStream}.
+ */
+public class SeekableAlluxioFileOutStreamIntegrationTest
```
Comment (LuQQiu) [inline tests/src/test/java/alluxio/client/fs/io/SeekableAlluxioFileOutStreamIntegrationTest.java:30]: Only write(int b) is tested, can you add tests for write(byte b[], int off, int len)?

**Review by LuQQiu — COMMENTED** [review]
Left some comments, the major ones are
how mPos be updated in write(byte b[], int off, int len) and getattr() when output stream with open mode

File: tests/src/test/java/alluxio/client/fs/io/SeekableAlluxioFileOutStreamIntegrationTest.java (hunk 2 above)
Comment (apc999) [reply tests/src/test/java/alluxio/client/fs/io/SeekableAlluxioFileOutStreamIntegrationTest.java:32]: added

**Review by apc999 — COMMENTED** [review]
replied

File: core/client/fs/src/main/java/alluxio/client/file/SeekableAlluxioFileOutStream.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+  @Override
+  public void write(byte b[]) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte b[], int off, int len) throws IOException {
+    mLocalFile.write(b, off, len);
+    mPos += len;
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/SeekableAlluxioFileOutStream.java:138]: Still not quite sure of this mPos += len
isn't it should be mPos = off + len?

**Pull request closed without merging**